### PR TITLE
COR-687: deduplication

### PR DESCRIPTION
**Pull request by internetti**
checking for duplicates in the database and within the uploaded file on request.

This is the query that should be build for 
Person 1: firstName1, middleName1, lastName1, id1, id2, id3
Person 2: firstName2, middleName2, lastName2, id4, id5, id6

```
SELECT * FROM individual_registrations 
WHERE country_id = 'countryID' 
AND deleted_at IS NULL 
AND 
  (
      identification_number_1 IN ('id1','id2','id3','id4','id5','id6')
      OR
      identification_number_2 IN ('id1','id2','id3','id4','id5','id6')
      OR
      identification_number_3 IN (SELECT * FROM UNNEST ('id1','id2','id3','id4','id5','id6'))
  )
AND 
  (
      (first_name = 'firstname1' AND middle_name = 'middleName1' AND last_name = 'lastName1')
      OR
      (first_name = 'firstName2' AND middle_name = 'middleName2' AND last_name = 'lastName2')
  )


SELECT * FROM individual_registrations 
WHERE country_id = 'countryID' 
AND deleted_at IS NULL 
AND 
  (
      identification_number_1 IN (SELECT * FROM UNNEST ($1::text[]))
      OR
      identification_number_2 IN (SELECT * FROM UNNEST ($1::text[]))
      OR
      identification_number_3 IN (SELECT * FROM UNNEST ($1::text[]))
  )
AND 
  (
      (first_name = $2 AND middle_name = $3 AND last_name = $4)
      OR
      (first_name = $5 AND middle_name = $6 AND last_name = $7)
  )

```

note: the order of key-value pairs and parameter indices can vary

### CODE REVIEW

File: internal/db/deduplicationOptions.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+			Columns:   []string{constants.DBColumnIndividualIdentificationNumber1, constants.DBColumnIndividualIdentificationNumber2, constants.DBColumnIndividualIdentificationNumber3},
+			Condition: LOGICAL_OPERATOR_OR,
+		},
+	},
+	DeduplicationOptionNameNames: {
+		ID:      DeduplicationOptionNameNames,
+		Label:   "Names",
+		Default: false,
+		Value: DeduplicationOptionValue{
+			Columns:   []string{constants.DBColumnIndividualFirstName, constants.DBColumnIndividualMiddleName, constants.DBColumnIndividualLastName},
```
Comment (internetti) [inline internal/db/deduplicationOptions.go:72]: might need to be updated to include native name?

File: pkg/views/forms/forms.go [hunk 2]
```diff
@@ -139,10 +139,10 @@ func computeStructFields(i interface{}) map[string]reflect.StructField {
 	for i := 0; i < fieldCount; i++ {
 		field := reflectType.Elem().Field(i)
 		fieldName := field.Name
-		jsonTag := field.Tag.Get("json")
-		if jsonTag != "" {
-			jsonTag = strings.Split(jsonTag, ",")[0]
-			fieldName = jsonTag
+		dbField := field.Tag.Get("db")
```
Comment (internetti) [inline pkg/views/forms/forms.go:142]: had to use this mapping also with the deduplication options, where I only have the db column names. there was no mapping from json tag to db column, it was more convenient this way

Comment (neb42) [reply pkg/views/forms/forms.go:142]: Maybe this should be:

```
tag := field.Tag.Get("json")
if tag == "" {
  tag := field.Tag.Get("db")
}
if tag != "" {
  .....
}
```

Comment (internetti) [reply pkg/views/forms/forms.go:142]: hm, if anything, I think it should be the same for all fields, not on a field by field basis

File: internal/api/individual_tabular.go [hunk 3]
```diff
@@ -766,91 +788,47 @@ func MarshalIndividualsExcel(w io.Writer, individuals []*Individual) error {
...
-		return time.RFC3339
-	case constants.DBColumnIndividualCreatedAt:
-		return time.RFC3339
-	case constants.DBColumnIndividualDeletedAt:
-		return time.RFC3339
-	default:
-		return "2006-01-02"
-	}
-}
+func FindDuplicatesInUpload(columns []string, records [][]string) containers.Set[string] {
```
[the remark this answers is not in the document]
Comment (internetti) [reply internal/api/individual_tabular.go:791]: yeah, the LoadStructs was a nightmare to work with, the structs were not allowed to have pointers in them, would have required some serious workarounds

File: web/templates/nav.gohtml [hunk 4]
```diff
@@ -1,176 +1,226 @@
...
+                                </div>
+
+                                <div class="form-check mt-5">
+                                    <input class="form-check-input"
+                                           type="checkbox"
+                                           value="true"
+                                           name="fileDeduplication">
+                                    <label class="form-check-label">
+                                        Also check for duplicates within the file
+                                    </label>
```
[the remark this answers is not in the document]
Comment (internetti) [reply web/templates/nav.gohtml:194]: this was talked about in the product teams chat. but this was before we introduced the upload limit.

File: web/templates/individual.gohtml [hunk 5]
```diff
@@ -192,4 +198,43 @@
         </div>
     </div>
     <!-- END DELETE CONFIRMATION MODAL -->
+
+    <!-- DEDUPLICATION MODAL -->
```
[the remark this answers is not in the document]
Comment (internetti) [reply web/templates/individual.gohtml:202]: I'm not convinced it would be worth it. I think it would end up being only the text and the checkboxes

**Comment by neb42** [comment]
> checking for duplicates in the database and within the uploaded file on request.
> 
> This is the query that should be build for Person 1: firstName1, middleName1, lastName1, id1, id2, id3 Person 2: firstName2, middleName2, lastName2, id4, id5, id6
> 
> ```
> SELECT * FROM individual_registrations 
> WHERE country_id = 'countryID' 
> AND deleted_at IS NULL 
> AND 
>   (
>       identification_number_1 IN ('id1','id2','id3','id4','id5','id6')
>       OR
>       identification_number_2 IN ('id1','id2','id3','id4','id5','id6'')
>       OR
>       identification_number_3 IN ('id1','id2','id3','id4','id5','id6'')
>   )
> AND 
>   (
>       (first_name = 'firstname1' AND middle_name = 'middleName1' AND last_name = 'lastName1')
>       OR
>       (first_name = 'firstName2' AND middle_name = 'middleName2' AND last_name = 'lastName2')
>   )
> ```
> 
> note: the order of key-value pairs can vary

I don't think the query is quite right. I think it should be:
```
SELECT * FROM individual_registrations 
WHERE country_id = 'countryID' 
AND deleted_at IS NULL 
AND 
  (
      identification_number_1 IN ('id1','id2','id3','id4','id5','id6')
      OR
      identification_number_2 IN ('id1','id2','id3','id4','id5','id6'')
      OR
      identification_number_3 IN ('id1','id2','id3','id4','id5','id6'')
      OR
      (first_name = 'firstname1' AND middle_name = 'middleName1' AND last_name = 'lastName1')
      OR
      (first_name = 'firstName2' AND middle_name = 'middleName2' AND last_name = 'lastName2')
  )
```

**Review by neb42 — COMMENTED** [review]
I still need to test it

File: internal/api/individual_deduplication.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+
+	duplicateScores := []containers.Set[int]{}
+	for i := 0; i < df.Nrow(); i++ {
+		duplicateScores = append(duplicateScores, containers.NewSet[int]())
+		GetDuplicationScoresForRecord(optionNames, df, i, duplicateScores[i])
+	}
+	return duplicateScores
+}
+
+func GetDuplicationScoresForRecord(optionNames []deduplication.DeduplicationTypeName, df dataframe.DataFrame, currentIndex int, duplicates containers.Set[int]) {
```
[the remark this answers is not in the document]
Comment (internetti) [reply internal/api/individual_deduplication.go:29]: how can I test it then?

[the remark this answers is not in the document]
Comment (internetti) [reply internal/api/individual_deduplication.go:29]: we need to check all tthe criteria (IDs, Names, Emails, whatever was chosen) before we can officially declare something a duplicate. 
so we go through all the criteria first, checking off every one that is fulfilled, and if at the end all criteria for a record are checked off, only then can we know that it is a duplicate.

File: web/templates/individual.gohtml [hunk 7]
```diff
@@ -192,4 +197,42 @@
...
+    <div class="modal" id="saveModal" tabindex="-1" role="dialog" aria-hidden="true">
+        <div class="modal-dialog" role="document">
+            <div class="modal-content">
+                <div class="modal-header"><h5>{{if .Individual.ID}}Update{{else}}Save{{end}}</h5></div>
+                <div class="modal-body">
+                    <div class="d-block">
+                        <p>
+                            If you want to prevent duplicate participants from being uploaded,
+                            please pick one or more of the criteria, so we know how to recognize duplicates.
```
Comment (internetti) [inline web/templates/individual.gohtml:210]: TODO: Update this text to let them know that the criteria will be connected by AND

**Comment by neb42** [comment]
File upload modal submit immediately after selecting the file instead of waiting for the submit button

**Comment by neb42** [comment]
Blank values aren't being treated properly. I haven't fully diagnosed the issue yet. Possibly they are ignored in the file check, but not the database check. I think they should be ignored everywhere

**Comment by internetti** [comment]
the issue with the blank values should be fixed, the upload input behaviour too.
not sure what you mean with the free fields, they behave like I'd expect them to

**Comment by neb42** [comment]
> the issue with the blank values should be fixed, the upload input behaviour too. not sure what you mean with the free fields, they behave like I'd expect them to

Yeah, sorry I meant that's expected. Just highlighting it's weird that they were different

**Comment by neb42** [comment]
Modal needs a cancel button

And there are sonar errors

**Comment by internetti** [comment]
> Modal needs a cancel button
> 
> And there are sonar errors

the sonar errors are about duplications of string variables in a test file, I'd ignore that. I tried to reduce soem code smell though.

**Review by neb42 — COMMENTED** [review]
Some minor comments but can merge after that.

There's a lot of codeql issues that should be addressed. I think we should merge this as it is pressing, but in future we shouldn't ignore codeql.

File: internal/db/migrations/postgres/030_update_individual_soft_delete_trigger.up.sql
[the remark this answers is not in the document]
Comment (internetti) [reply internal/db/migrations/postgres/030_update_individual_soft_delete_trigger.up.sql:1]: line 37, I added an ID to the error message, to help debugging

File: cmd/mockdata/main.go
[the remark this answers is not in the document]
Comment (internetti) [reply cmd/mockdata/main.go:1]: yes, will remove it from here

**Review by neb42 — APPROVED** [review]

**Pull request merged**